airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,407 @@
+
+    stream_slice_period: Mapping[str, int] = {"days": 1}  # max period is 31 days (API limit)
+
+    requests_per_minute: int = 30  # API limit is 50 reqs/min from 1 IP to all endpoints, otherwise IP is banned for 5 mins
+
+    def __init__(
+        self,
+        start_date: Optional[datetime] = None,
+        end_date: Optional[datetime] = None,
+        is_sandbox: bool = False,
+        config: Optional[Mapping[str, Any]] = None,
+        **kwargs,
+    ):
+
+        # Initiate data from config
+        if config:
+            start_date = isoparse(config["start_date"])
+            end_date = isoparse(config.get("end_date")) if config.get("end_date") else None
+            is_sandbox = config["is_sandbox"]
+
+        self.start_date = start_date
+        self.end_date = end_date
+        self.is_sandbox = is_sandbox
+
+        self._validate_input_dates()
+
+        super().__init__(**kwargs)
+
+    @property
+    def end_date(self):
+        """Return initiated end_date or now()"""
+        now = datetime.now().replace(microsecond=0).astimezone()
+        if not self._end_date or self._end_date > now:
+            # If no end_date or end_date is in future then return now:
+            return now
+        else:
+            return self._end_date
+
+    @end_date.setter
+    def end_date(self, value: datetime):
+        self._end_date = value
+
+    def _validate_input_dates(self):
+
+        # Validate input dates
+        if self.start_date > self.end_date:
+            raise Exception(f"start_date {self.start_date.isoformat()} is greater than end_date {self.end_date.isoformat()}")
+
+        current_date = datetime.now().replace(microsecond=0).astimezone()
+        current_date_delta = current_date - self.start_date
+
+        # Check for minimal possible start_date
+        if current_date_delta > timedelta(**self.start_date_min):
+            raise Exception(
+                f"Start_date {self.start_date.isoformat()} is too old. "
+                f"Min date limit is {self.start_date_min} before now:{current_date.isoformat()}."
+            )
+
+        # Check for maximum possible start_date
+        if current_date_delta < timedelta(**self.start_date_max):
+            raise Exception(
+                f"Start_date {self.start_date.isoformat()} is too close to now. "
+                f"Max date limit is {self.start_date_max} before now:{current_date.isoformat()}."
+            )
+
+    @property
+    def url_base(self) -> str:
+
+        return f"{get_endpoint(self.is_sandbox)}/v1/reporting/"
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+
+        return {"Content-Type": "application/json"}
+
+    def backoff_time(self, response: requests.Response) -> Optional[float]:
+        # API limit is 50 reqs/min from 1 IP to all endpoints, otherwise IP is banned for 5 mins
+        return 5 * 60.1
```

Comment:
So, API is blocked for  5 mins, what if some API is blocked for 30 mins or more, should we still handle this by such custom backoff, or maybe it's better to exit with error?

Reply:
That's a good question. In that case we should probably fail for now.